### ISSUE TITLE
LLVM 4.0 support

### ISSUE THREAD

**Issue by chneukirchen** [Morgan-Stanley/hobbes#28]
hobbes currently does not build with LLVM 4.0 out of the box, but only needs slight patching.

Biggest issue is that most ifdef only compare LLVM_VERSION_MINOR, I'd recommend to add a macro like LLVM_VERSION_ATLEAST(major,minor). Then it's essentially enough to use the newest ifdef for LLVM 4.0, and:

In jitcc.H add
```
#include "llvm/IR/IRBuilder.h"
#include "llvm/IR/LegacyPassManager.h
```

in llvm.H, replace include for llvm/Bitcode/ReaderWriter.h with llvm/Bitcode/BitcodeReader.h and llvm/Bitcode/BitcodeWriter.h, and rewrite context() to return a static instance of llvm::LLVMContext.

I have attached a quick patch that makes it build and pass the test suite, if you want I can prepare a real PR, but perhaps you may want to tweak this anyway.

[0001-llvm-4.0-support.patch.txt](https://github.com/Morgan-Stanley/hobbes/files/1155628/0001-llvm-4.0-support.patch.txt)

There is another problem, but it may be due to how packaging works on my distribution: it links all `.a` files of LLVM and the `libLLVM-4.0.so`, which results in duplicate instance of certain objects which break LLVM immediately. I haven't found a better way than editing the `link.txt` files manually to fix this yet. (`llvm-config --libs` correctly prints `-lLLVM-4.0`, not sure what cmake does.)

**Comment by kthielen** [comment]
Thank you!  I would love to take a PR -- there's a form that MS wants to have signed/scanned by folks prior to accepting a PR.  I could send the form to you, but otherwise I will take a look at this later and keep this issue open until hobbes works with LLVM 4.0.

**Comment by chneukirchen** [comment]
I think someone familiar with the code base should be able to take a look at that patch and do it properly easily, no paper work needed.

**Comment by kthielen** [comment]
It looks like some of these changes don't work with earlier LLVM builds enabled, I will look into those issues and find a way to adjust so that it works for all.  Until it's resolved though, I will leave the issue open.

Thank you for your contribution!

**Comment by george-ball** [comment]
I can't get it to build out of the box on MacOS 10.12, the macros seem to expect LLVM 3.5 and nothing else, and this version is not available on homebrew... So would welcome this fix :)

**Comment by chneukirchen** [comment]
Ah yes, of course the include changes will need to be #ifdef guarded.

**Comment by kthielen** [comment]
George are you referring to the current state of master or with this change for LLVM 4.0?  Our pull request continuous build tool checks that it builds for OSX with LLVM 3.7 and I've built it with 3.5 and 3.7 locally (with macOS 10.12.5).

**Comment by george-ball** [comment]
OK Let me try again, I thought I was running with the latest version from GH, could be finger trouble on my part tho’

G

**Comment by george-ball** [comment]
My apologies, now built successfully and was my fault!
G

**Comment by pritambaral** [comment]
The issue with cmake generating linkages with both 'all `.a` files' and `libLLVM-4.0.so` is because of the `all` in https://github.com/Morgan-Stanley/hobbes/pull/5#issuecomment-311241354. This patch with the selective components from before that PR builds and runs properly with LLVM4.0 on Linux.

**Comment by kthielen** [comment]
Yeah the 'all' selection is a little gross right?  I knew it at the time, but couldn't seem to get the right component selection across different versions of LLVM.

**Comment by pritambaral** [comment]
Was there any contention except `jit` (<= 3.5) vs. `mcjit` (> 3.5)?

BTW, Using the component selection from before that patch gives a significantly smaller (by 90%) binary. Static linking doesn't prune linked libraries.

**Comment by kthielen** [comment]
It wasn't contention so much as coming up with a list that would work for all versions of LLVM that people might select.  Or maybe that's what you mean (sorry if I'm misunderstanding).

Completely agree, it's gross to pick "all" and bloat executables with this.

**Comment by pritambaral** [comment]
I am able to build with 4.0 and 3.5 with the component selection as fine as the original one; the only difference in components required between the two version is `jit` vs `mcjit` (the `passes` component isn't strictly required, it's actually the older `ipo` component that is needed). Such a difference is easily handled in a single if-else. 

There were other minor changes required in some source files to get it to build with both 3.5 and 4.0. I haven't tested against 3.7 (my distro doesn't package it) and the build breaks mysteriously with 3.9 (portions of the code laboriously check for each individual minor version number instead of a range, and I found at least one that stopped at `LLVM_VERSION MINOR = 8`.

**Comment by kthielen** [comment]
What does the easy single if-else look like?  It would be great to make that component selection based on the LLVM version that we're building against (without requiring new build arguments).

I have a local change building with LLVM 4.0 on OSX but I'm running into a new error when running tests.  I will have to get some time to work through it.  Hopefully can get it soon.

Yes you're right, the logic for dealing with different LLVM versions is a little disjointed, probably expediently selected.  It would be a good idea to simplify it at some point.

**Comment by pritambaral** [comment]
```CMakeLists.txt
if (${LLVM_PACKAGE_VERSION} VERSION_LESS "3.6")
  set(jit_lib jit)
else()
  set(jit_lib mcjit)
endif()
# 'x86' module should have covered 'x86asmparser' and 'x86codegen' already, but using 'x86' alone didn't work
llvm_map_components_to_libnames(llvm_libs x86asmparser x86codegen x86 ipo ${jit_lib})
```

**Comment by kthielen** [comment]
Thank you!  I've updated the llvm-4.0 PR (referenced above) with your method.  I tested it locally on an OSX machine with LLVM 3.5 and 4.0.  It seems that the final compiled executables are the same size though, FWIW.

**Comment by pritambaral** [comment]
> It seems that the final compiled executables are the same size though, FWIW.

How big are they? A good build should come out in the 1-2M range.

Of course, what gets linked in still depends on your LLVM build. My distro packages LLVM as a single-file shared library, and split statically-linkable archives. If one's build has only a single-file statically-linkable archive, then, of course, the final executable is going to be large.

**Comment by kthielen** [comment]
Still ~20MB for me, but if this gets you a reasonable executable size then it's great progress.  In the long run it would be good to minimize executable size.

**Comment by pritambaral** [comment]
> 20MB

Wow. The build on Arch Linux with `all` components was 13M per binary. I think your size is large because of your LLVM build. With finely chosen components the build on Arch Linux is 1.2M. You could have a look in your `link.txt` files to see what gets linked in in your build.

If this is indeed a case of LLVM build not having separate link archives, then the "long run" is already here.

**Comment by skaller** [comment]
Ok, i have built LLVM 4.01 and applied fixed it (partially with the patches in the OP). Now I can compile but the link fails missing typeinfo().

```
[ 42%] Linking CXX executable hi
Undefined symbols for architecture x86_64:
  "typeinfo for llvm::CallbackVH", referenced from:
....
```

I have no idea how this can work. Despite the incorrect notes (requires g++) in the README, the cmake build process is using the Apple XCode version of Clang and presumably its LLVM as well. Setting LLVM_REQUIRES_RTTT=1 during the make has no effect. OSX 10.12.1.

**Comment by kthielen** [comment]
README issue noted, thank you.  We will update the doc to reflect the fact that we are building hobbes on OSX and Linux, with g++ and clang.

We do have hobbes building with tests passing for OSX 10.12 (macOS Sierra) with LLVM 3.5, 3.7, 4.0 (but the 4.0 support is new with the PR I linked above, not yet merged to master).  As discussed in a different issue, hobbes uses RTTI and at least for LLVM 3.7+ I believe this means that LLVM also has to be built with RTTI (due to the way that LLVM interfaces must be implemented).

Most folks who are running hobbes on OSX have installed LLVM through brew or ports and have it working correctly.  Our CI agent does this as well, since all PRs are gated on successful builds and tests on Linux and OSX.

**Comment by skaller** [comment]
I don't use brew. (package managers on OSX are dangerous because they can duplicate other installs including Apples leading to confusion)

I downloaded and built LLVM 4.01 (the latest) by hand. XCode of course does come with LLVM. I have no idea which version. However some notes suggest the cmake for LLVM is required which doesn't seem quite right (is it really installed in XCode?). However Xcode doesn't come with the LLVM assembler. It's really unclear when you just say "requires LLVM" what is actually required. Situation is not helped by the abysmal documentation on the LLVM web site.

So I'd appreciate step by step instructions. LLVM 4.01 is installed in /usr/local/*. The system LLVM used by clang is in the usual place. Clang will use the system (XCode) one, not the one I installed. I can rebuild LLVM if required. I prefer NOT to have to also build the clang which goes with it, that's a nightmare, but the build notes don't say anything about requiring the matching clang :)

The standard clang++ provides typeinfo() without any switches, this works fine:

```
#include <typeinfo>
#include <iostream>
#include <string>
using namespace std;

int main()
{
  auto x = &typeid(string("Hello"));
  cout << x -> name() << endl;
}
```
Its a real pain XCode doesn't include the LLVM assembler. GHC can build to LLVM too, but you need the assembler.

**Comment by kthielen** [comment]
The LLVM you link to must have been built with RTTI enabled.  You can build LLVM without a package manager, but for that I think that the best source for step-by-step instructions is (or should be?) the LLVM page itself.  If you don't think their docs are adequate but you do get this working, I would love to put up your build instructions.  I agree that the "build" instructions in the hobbes README should be updated and expanded.

The problem is that LLVM provides some interfaces which applications must inherit from in order to implement some basic functions in the JIT compiler.  When a hobbes class inherits from an LLVM class, if the hobbes class requires RTTI then RTTI is also required of the LLVM class.  However if the LLVM library you link to was not built with RTTI enabled, then you will get a linker error like the one you found.

**Comment by skaller** [comment]
The problem is your build script is broken. It cannot be MY problem because I have no idea what your build script is doing. It provides no output, no log that I know of. If you require a feature, you should test for it. Once the build starts it should work. *I* am not "linking to LLVM". Its YOUR script which is doing that. I have no idea whatsoever what it links to. I cannot read any of that stuff. I can look at your system once it is built, i can look at the C++, but if it doesn't build "out of the box" it is necessarily only a problem you can solve. I built an LLVM. I rebuilt it with LLVM_REQUIRES_RTTI=1 as well. It's installed in /usr/local. I have no idea how it can *possibly* be used in a C++ compile using the XCode Clang which is forcibly connected to the XCode LLVM which definitely has RTTI enabled. However if you're ALSO running LLVM assembler it is going to fine the /usr/local/bin executable. Which is NOT the same one that Clang is using. Since we're just trying to link the first executable in your script, I guess there's no assembler involved. So it should work. My test C++ code works. XCode LLVM has RTTI.

I've spent weeks helping users of my system get a build running. The latest saga is that Travis is using a broken g++ so the binaries it produces are useless on later Ubuntu versions. It's not their fault, its the fault of the packagers doing the backports.

**Comment by skaller** [comment]
Ok so, with make -d, I can see its trying to think to my /usr/local/ version of LLVM. Clang itself is using the Xcode LLVM to make object files, but then we're linking them to a different LLVM which hopefully should be OK. Starting from scratch, rebuilding LLVM now. Unfortunate cmake feature, even with -d on the make, the commands being executed are not displayed. I get pretty colour and a summary line useless for solving problems. (the system I use does the same thing, but it logs the actual commands to a file).

**Comment by kthielen** [comment]
OK, I am willing to accept that the build process can be improved, definitely not fighting you on that.  TBH the cmake infra was introduced as an alternative to the internal build process we use at MS and so I'm still learning too.  Please be patient with me.  :)

When I have built with cmake, I've always had to give the LLVM_DIR to a particular LLVM version's cmake files.  I like your idea to detect necessary features from the selected LLVM version at cmake-time and block makefile generation if it's not going to work.  I agree that it's very frustrating to get through most of the process just to find that the link step won't work and then have to backtrack.  I will have to research more about cmake and the cmake setup in LLVM but if you have pointers I would gladly accept/merge them (as I did above for pritambaral's suggestion which has now been merged to master).

**Comment by skaller** [comment]
Sorry, I have fought with my own build system for a very long time. I find, less than half the OSS projects out there actually build. Part of the problem is that ALL of them use totally the wrong design .. except the one I designed. Goal oriented building is totally wrong. Building has to be bottom up not top down. 

Anyhow, I have rebuilt LLVM:

```
~/Desktop/llvm>cmake -DLLVM_ENABLE_EH=ON -DLLVM_ENABLE_RTTI=ON  ../llvm-4.0.1.src
```

and now hobbes appears to build! There is a minor problem which I get in my system in one case, you get it too:

```
ld: warning: direct access in function 'llvm::DOTGraphTraits<llvm::SelectionDAG*>::getNodeAttributes(llvm::SDNode const*, llvm::SelectionDAG const*)' from file '/usr/local/lib/libLLVMSelectionDAG.a(SelectionDAGPrinter.cpp.o)' to global weak symbol 'std::__1::char_traits<char>::eq(char, char)' from file 'CMakeFiles/hog.dir/bin/hog/batchrecv.C.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```

I get that for multiple references. I think it is a bug in libc++.

I cannot see how you are doing the build, its hidden inside incomprehensible make stuff.
The situation here is you have to build shared library code with -visibility=hidden,
and then override the symbols you want to export in the C++ source with a special
attribute. I think the Clang people messed up in a couple of cases. I put a bug on their
bug tracker but have heard nothing. I also get a bug in a string related class.
It has something to do with the disaster of C++ strings, where the ISO committee
"clarified" that Copy on Write is not allowed (it isn't thread safe). Both GNU and Clang
had to add new conforming implementation and keep the old one too for compatibility.
The one you see above with "std::__1::" prefix is the new (conforming) one.
[On Travis, you get the old non-conforming one no matter what you do because
the g++ backport is screwed up]

**Comment by kthielen** [comment]
Thanks for the info about ld warnings and visibility.  I don't think that the hobbes build should make any .so/.dylib files but it does make a libhobbes-pic.a with -fPIC to be used in .so projects (we have a few such projects internally).  I wonder if that's where the warnings come from -- I haven't seen them before.

I agree with you about the opaque cmake-generated build hiding a lot of information behind its output, not very useful when/if something goes wrong.

I'm interested in your ideas about building.  It sounds like you've thought about this problem a lot and have recognized common mistakes across other OSS projects as well.  I've taken dependencies out of this project in the past to make building easier (e.g.: hobbes used to depend on boost for a few things).  It might be an extreme step to take out LLVM for that reason, but anything that can be done to simplify the build process will help a lot of people.

**Comment by skaller** [comment]
I'm surprised you haven't see those warnings.. are we not building with the same Xcode on the same OS version? OSX 10.12.1 (Sierra).

I am thinking about embedding hobbes in Felix. I can do this because Felix is a C++ code generator and you can embed hobbes in C++, and I can embed C++ in Felix :-)

Now the thing is, Felix generates libraries (a program is a library that happens to have side effects from global initialisations). And the default is shared library. Which means I prefer to use hobbes and llvm etc as shared libraries, so I'd be wanting to convert libhobbes_pic.a into libhobbes_pic.dylib (.so on Linux, .dll on Windows).

My system can do static linkage as well, but it autobuilds everything, and there is no option to mix static and dynamic linkage of libraries. (Its generally a bad idea anyhow).

Glancing at the code, you have no visibility annotations in the C++ which means you have to compile with visibility=default which exports all extern symbols (like with static linkage) but is considered archaic. It's also a *major* hassle with C++ due to automatic template instantiation: what happens if two shared libraries use the same specialisation of a template and it is exported? Then you have duplicate symbols. Ditto for typeinfo, exceptions, dynamic casts.

There's a special linkage mode to handle that (vague linkage I think its called) which is, of course a disaster.

That's basically where the warnings are coming from. A conflict in visibility modes.

**Comment by kthielen** [comment]
Closing this issue as LLVM 4.0 support has been merged to master, a few follow up changes will also soon be in master with PR https://github.com/Morgan-Stanley/hobbes/pull/38
